### PR TITLE
Sprint 21 gw

### PR DESCRIPTION
**Pull request by G-White-ISB**
Support to update cohort versions (https://app.zenhub.com/workspaces/overall-project-workspace-5d39f406a4f118000112b4bb/issues/imagingdatacommons/idc-projectmanagement/956). Older cohorts appear in the cohort list with a 'compare' button. Pressing the button creates a dialog window showing the number of cases, studies,series for the old cohort and a new cohort created by apply the old cohort filter to the new data version.  Pressing a button in the dialog window opens the explorer page using the old cohort's filter to create a new cohort with the new data version. The user must then save the new cohort.

### CODE REVIEW

**Comment by G-White-ISB** [comment]
There is some work on refactoring code for populating the tables. This is not finished but the new code is not being used yet (its an inactive 'view')

File: idc/views.py [hunk 1]
```diff
@@ -240,6 +240,25 @@ def help_page(request):
 def quota_page(request):
     return render(request, 'idc/quota.html', {'request': request, 'quota': settings.IMG_QUOTA})
 
+@login_required
```
Comment (s-paquette) [inline idc/views.py:243]: Not clear on what this does?

File: idc/views.py [hunk 2]
```diff
@@ -240,6 +240,25 @@ def help_page(request):
...
+def fetch_cohort_data(request):
```
Comment (s-paquette) [inline idc/views.py:244]: Not clear on what this view does?

File: idc/views.py [hunk 3]
```diff
@@ -264,11 +284,29 @@ def explore_data_page(request):
...
+                cohort_filters_list = cohort_filters_dict[0]['filters']
+                for cohort in cohort_filters_list:
+                    cohort_filters[cohort['name']] = cohort['values']
+
+
+        if wcohort and is_json:
+            filters = cohort_filters
         context = build_explorer_context(is_dicofdic, source, versions, filters, fields, order_docs, counts_only,
-                                         with_related, with_derived, collapse_on, is_json, uniques=uniques)
+                                         with_related, with_derived, collapse_on, is_json, uniques=uniques, totals=totals)
```
Comment (s-paquette) [inline idc/views.py:309]: There's no need to ask for totals in the explorer context call; they're saved in the cohort. You can do it above (where indicated).

File: idc/views.py [hunk 4]
```diff
@@ -264,11 +284,29 @@ def explore_data_page(request):
...
         record_limit = int(req.get('record_limit', '2000'))
         offset = int(req.get('offset', '0'))
-
+        #sort_on = req.get('sort_on', collapse_on+' asc')
+        cohort_id = req.get('cohort_id','-1')
+
+        cohort_filters={}
+        if (int(cohort_id)>-1):
+            cohort = Cohort.objects.get(id=cohort_id, active=True)
+            cohort.perm = cohort.get_perm(request)
```
Comment (s-paquette) [inline idc/views.py:297]: You can ask for the case, study, and series count here; the names of the attributes are:
- case_count
- series_count
- study_count
- collections: This is a string list of the collections with data in this cohort

**Review by s-paquette — CHANGES_REQUESTED** [review]
Just a couple of small things to adjust, otherwise looks good!

File: idc/views.py (hunk 3 above)
Comment (G-White-ISB) [reply idc/views.py:309]: But right now I am reporting the totals  by applying the old filter on the new data version BEFORE the new cohort is actually created. Example :clicking the 'Compare' button on cohort 4 on the /cohorts/ page opens this dialog which shows totals for the actual cohort 4 and those of the cohort that would be created by applying the filter to the current data version. But that new cohort doesn't exist yet. Clicking the button 'Load New Version' just goes to the explorer page with that filter preselected. The user still has to save the cohort. Btw ignore those zeros in screenshot - they're due to my issue running the stats script.
Screen Shot 2021-04-30 at 5 03 01 PM

File: idc/views.py (hunk 2 above)
Comment (G-White-ISB) [reply idc/views.py]: An aborted refactor. Sorry I missed this. I just deleted it.

File: idc/views.py (hunk 3 above)
Comment (s-paquette) [reply idc/views.py:291]: For that you definitely don't want to run the entire context builder, as it results in a large amount of unneeded data being pulled (records, etc.) explicitly for displaying the main page.

All you'll want to do there is use _get_cohort_stats(), sending in 0 for a cohort ID and then the cohort filters and the data sources for the new version. So something like:

```
from cohorts import _get_cohort_stats

new_version_stats = _get_cohort_stats(
   0,
   old_cohort.cohort.get_filters_as_dict()[0]['filters'], 
   ImagingDataCommons.objects.get(active=True).get_data_sources(active=True, source_type=DataSource.SOLR, aggregate_level="StudyInstanceUID")
)
```

No need for the explorer view--it's not intended to be run at any time except to actually load the full explorer page.

**Comment by G-White-ISB** [comment]
Refactored so that explorer view is not used in getting cohort stats. There did not seem to be a cohort stats view so I created one that can report saved stats or generate new stats using the cohort filter as you had suggested above. This view is in IDC-Common. There is a new PR in IDC-Common corresponding to this.

**Review by s-paquette — APPROVED** [review]

**Pull request merged**